doxygen/contributor_help_pages/require_meta.md

```diff
@@ -1,22 +1,263 @@
+set of function template specialization is untenable for many
+functions in the Math library. For example, a single argument of the
+function may take 7 distinct C++ types: `double`, `stan::math::var`,
+`std::vector<double>`, `std::vector<stan::math::var>`,
+`Eigen::Matrix<double, -1, 1>`, `Eigen::Matrix<stan::math::var, -1,
+1>`, or `stan::math::var_value<Eigen::Matrix<double, -1, -1>>`.
+For a 3-argument function, we would need to define 343 (7^3) different
+function template specializations to handle all the autodiff types.
+
+In the Math library, we use a technique that allows the definition of
+multiple template functions where each handles a subset of allowable
+types. We add a pointer [non-type template
+parameter](https://en.cppreference.com/w/cpp/language/template_parameters#Non-type_template_parameter)
+to the template parameter list with a default value of `nullptr`. Any
+`void*` non-type template parameter with a default of `nullptr` is
+valid and the non-type template parameter is ignored by the
+compiler. Utilizing [substitution failure is not an error
+(SFNIAE)](https://en.cppreference.com/w/cpp/language/sfinae), a
+substitution failure for the non-type template parameter will result
+in that definition being removed from the possible function
+definitions. Using this technique we have to be careful not to violate
+the [One Definition
+Rule](https://en.cppreference.com/w/cpp/language/definition) and only
+provide one definition for any set of types.
+
+For convenience in using this technique, the Math library has
+implemented a set of [`requires` type traits](@ref require_meta).
+When we pass a type that satisfies the `requires` type trait, the
+trait evaluates to `void`. When a type that does not satisfy the
+`requires` template parameter is passed, there is a substitution
+failure. These traits are used in the template functions by adding a
+call to a `requires` type trait to the parameter list.
+
+Below is an example to illustrate how this technique is used. After
+the example, the rest of this page describes what the requires type
+traits are, how to use them, and how to add new ones if necessary.
+
+#### Example
+
+Here's a function that would have two different template functions for
+`stan::math::var` and `double`:
 
-The [`requires` template parameters](@ref require_meta) type traits are aliases for `std::enable_if_t` that have premade conditions for turning on and off function definitions during compilation.
-These are useful for having generalized templates while still overloading a function or class.
-You can think of these as "legacy concepts".
-These are used in a very similar fashion to C++20's `requires` keyword.
+```
+template <typename T, requires_var_t<T>* = nullptr>
+T foo(const T& t) {
+  // handles stan::math::var
+  return t;
+}
+
+template <typename T, requires_not_var_t<T>* = nullptr>
+T foo(const T& t) {
+  // handles primitives
+  return t;
+}
```

Comment:
Think we should flip these to show "general version" and "overload for var"
```suggestion
template <typename T, requires_not_var_t<T>* = nullptr>
T foo(const T& t) {
  // handles primitives
  return t;
}

template <typename T, requires_var_t<T>* = nullptr>
T foo(const T& t) {
  // handles stan::math::var
  return t;
}
```

Reply:
That sounds good! (I had originally written that way too, so I think it's more natural to think that way, but then thought it was easier having the "not" condition second. If it's easier to understand that way, then great!)